numba_dpex/caching.py

```diff
@@ -0,0 +1,131 @@
+        #         "as it uses dynamic globals "
+        #         "(such as ctypes pointers and large global arrays)"
+        #     )
+        # if cannot_cache:
+        #     msg = 'Cannot cache compiled function "%s" %s' % (
+        #         cres.fndesc.qualname.split(".")[-1],
+        #         cannot_cache,
+        #     )
+        #     warnings.warn_explicit(
+        #         msg, NumbaWarning, self._locator._py_file, self._lineno
+        #     )
+        #     return False
+        return True
+
+
+class SpirvKernelCache(_Cache):
+    """
+    Implements a cache that saves and loads CUDA kernels and compile results.
+    """
+
+    _impl_class = SpirvKernelCacheImpl
+
+    def __init__(self, py_func):
+        self._name = repr(py_func)
+        self._py_func = py_func
+        self._impl = self._impl_class(py_func)
+        self._cache_path = self._impl.locator.get_cache_path()
+        # This may be a bit strict but avoids us maintaining a magic number
+        source_stamp = self._impl.locator.get_source_stamp()
+        filename_base = self._impl.filename_base
+        self._cache_file = IndexDataCacheFile(
+            cache_path=self._cache_path,
+            filename_base=filename_base,
+            source_stamp=source_stamp,
+        )
+        self.enable()
+
+    @property
+    def cache_path(self):
+        return self._cache_path
+
+    def enable(self):
+        self._enabled = True
+
+    def disable(self):
+        self._enabled = False
+
+    def flush(self):
+        self._cache_file.flush()
+
+    def load_overload(self, sig, target_context):
+        if not self._enabled:
+            return
+        key = self._index_key(sig, target_context.codegen())
+        data = self._cache_file.load(key)
+        return data
+
+    def save_overload(self, sig, data, target_context):
+        if not self._enabled:
+            return
+        if not self._impl.check_cachable(data):
+            return
+        self._impl.locator.ensure_cache_path()
+        key = self._index_key(sig, target_context.codegen())
+        data = self._impl.reduce(data)
+        self._cache_file.save(key, data)
+
+    def _index_key(self, sig, codegen):
+        """
+        Compute index key for the given signature and codegen.
+        It includes a description of the OS, target architecture and hashes of
+        the bytecode for the function and, if the function has a __closure__,
+        a hash of the cell_contents.
+        """
+        codebytes = self._py_func.__code__.co_code
+        if self._py_func.__closure__ is not None:
+            cvars = tuple([x.cell_contents for x in self._py_func.__closure__])
+            # Note: cloudpickle serializes a function differently depending
+            #       on how the process is launched; e.g. multiprocessing.Process
+            cvarbytes = dumps(cvars)
+        else:
+            cvarbytes = b""
+
+        return (
+            sig,
+            codegen.magic_tuple(),
```

Comment:
The key needs to incorporate the `device_type` and the `backend`. Otherwise we may cache a kernel compiled for a GPU and try to run it on a different type of a device. Look at the current logic in `numba_dpex/compiler.py`